### ISSUE TITLE
onComplete/onNext for observable (sometimes) not getting called after async write

### ISSUE THREAD

**Issue by curiousily** [realm/realm-java#2978]
#### Goal

Create/update realm objects and immediately find the saved results

#### Expected Results

Find all saved (updated or created) realm objects

#### Actual Results

onComplete of the observable is called only for 7-9/10 cases

#### Steps & Code to Reproduce

Please, see the sample code below. In about every 1-3 calls out of 10 the onNext/onComplete is never get called. Here is a sample output:

```
06-10 12:34:00.764 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:00.765 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 0 total completes: 1
06-10 12:34:01.690 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:01.690 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 1 total completes: 2
06-10 12:34:02.846 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:02.847 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 3 total completes: 3
06-10 12:34:04.053 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:04.053 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 4 total completes: 4
06-10 12:34:05.082 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:05.082 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 6 total completes: 5
06-10 12:34:06.685 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:06.685 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 8 total completes: 6
06-10 12:34:07.871 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:07.871 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 7 total completes: 7
06-10 12:34:08.939 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:08.939 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 9 total completes: 8
06-10 12:34:10.148 7561-7561/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:34:10.149 7561-7561/realmtest.com.realmtest D/RXTestRealm: onComplete: call 5 total completes: 9
```

As you can see from the output - the total calls are 9 (instead of 10) and call number 2 is missing.
The sample below is stripped version of a real project but still we're able to reproduce it with it.

#### Code Sample

```java

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private int totalCompletes;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.refresh_tasks).setOnClickListener(this);
        RealmConfiguration config = new RealmConfiguration.Builder(this)
                .schemaVersion(BuildConfig.VERSION_CODE)
                .deleteRealmIfMigrationNeeded()
                .build();
        Realm.setDefaultConfiguration(config);
        totalCompletes = 0;
    }

    private void refreshTasks(final int callNum) {
        List<Task> tasks = new ArrayList<>();
        for (int i = 0; i < 1000; i++) {
            tasks.add(new Task(UUID.randomUUID().toString(), UUID.randomUUID().toString()));
        }
        saveTasks(tasks).flatMap(new Func1<List<Task>, Observable<List<Task>>>() {
            @Override
            public Observable<List<Task>> call(List<Task> tasks) {
                for (Task t : tasks) {
                    t.setName(UUID.randomUUID().toString());
                }
                return saveTasks(tasks);
            }
        }).flatMap(new Func1<List<Task>, Observable<?>>() {
            @Override
            public Observable<List<Task>> call(List<Task> tasks) {
                return findTasks();
            }
        }).subscribe(new Action1<Object>() {
            @Override
            public void call(Object o) {
                Log.d("RXTestRealm", "onNext");
            }
        }, new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                Log.d("RXTestRealm", "onError " + throwable.getMessage());
            }
        }, new Action0() {
            @Override
            public void call() {
                totalCompletes++;
                Log.d("RXTestRealm", "onComplete: call " + callNum + " total completes: " + totalCompletes);
            }
        });
    }

    private Observable<List<Task>> findTasks() {
        final Realm realm = Realm.getDefaultInstance();
        return realm.where(Task.class).findAllAsync().asObservable().filter(new Func1<RealmResults<Task>, Boolean>() {
            @Override
            public Boolean call(RealmResults<Task> tasks) {
                return tasks.isLoaded();
            }
        }).first().map(new Func1<RealmResults<Task>, List<Task>>() {
            @Override
            public List<Task> call(RealmResults<Task> tasks) {
                List<Task> res = realm.copyFromRealm(tasks);
                realm.close();
                return res;
            }
        });
    }

    private Observable<List<Task>> saveTasks(final List<Task> tasks) {
        return Observable.create(new Observable.OnSubscribe<List<Task>>() {
            @Override
            public void call(final Subscriber<? super List<Task>> subscriber) {
                final Realm realm = Realm.getDefaultInstance();
                realm.executeTransactionAsync(new Realm.Transaction() {
                    @Override
                    public void execute(Realm backgroundRealm) {
                        backgroundRealm.copyToRealmOrUpdate(tasks);
                    }
                }, new Realm.Transaction.OnSuccess() {
                    @Override
                    public void onSuccess() {
                        realm.close();
                        subscriber.onNext(tasks);
                        subscriber.onCompleted();
                    }
                }, new Realm.Transaction.OnError() {
                    @Override
                    public void onError(Throwable error) {
                        realm.close();
                        subscriber.onError(error);
                    }
                });
            }
        });
    }

    @Override
    public void onClick(View v) {
        for(int i = 0; i < 10; i++) {
            refreshTasks(i);
        }
    }
}

```
Task class

```java

public class Task extends RealmObject {


    @PrimaryKey
    @Required
    private String id;

    public void setName(String name) {
        this.name = name;
    }

    private String name;

    public Task() {

    }

    public Task(String id, String name) {
        this.id = id;
        this.name = name;
    }

    public String getName() {
        return name;
    }
}
```

**Important**: It seems that using sync writes solves the issue

saving objects code:

```java
    final Realm realm = Realm.getDefaultInstance();
    realm.executeTransaction(new Realm.Transaction() {
                    @Override
                    public void execute(Realm realm) {
                        realm.copyToRealmOrUpdate(tasks);
                    }
                });
    realm.close();
    subscriber.onNext(tasks);
    subscriber.onCompleted();
```

Here the output is correct:

```
06-10 12:52:38.713 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:38.713 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 0 total completes: 1
06-10 12:52:39.718 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:39.718 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 1 total completes: 2
06-10 12:52:41.347 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:41.347 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 2 total completes: 3
06-10 12:52:42.553 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:42.553 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 7 total completes: 4
06-10 12:52:43.575 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:43.575 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 4 total completes: 5
06-10 12:52:44.716 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:44.717 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 3 total completes: 6
06-10 12:52:45.787 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:45.787 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 8 total completes: 7
06-10 12:52:47.530 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:47.531 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 9 total completes: 8
06-10 12:52:48.784 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:48.784 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 5 total completes: 9
06-10 12:52:49.804 20172-20172/realmtest.com.realmtest D/RXTestRealm: onNext
06-10 12:52:49.810 20172-20172/realmtest.com.realmtest D/RXTestRealm: onComplete: call 6 total completes: 10
```

#### Version of Realm and tooling
Realm version(s): 1.0.0

Android Studio version: 2.1.2

Which Android version and device: 5.1.1 (Nexus 4)

**Comment by stk1m1** [comment]
Hi @curiousily

Thanks for the detailed report. I've tried your sample and my outcome is quite irregular. Some times, I only recieve 2 calls `onComplete()`. I've followed your code and you're filtering object for its validity in `findTasks()` using another async call. May I ask your rational behind the logic please?

**Comment by stk1m1** [comment]
@curiousily 

I've checked the your sample again and I've realized that you're actually calling async query within an async transaction. I'd recommend to pair up `executeTransaction()` + `findAllAsync()` or `executeTransactionAsync()` + `findAll()` then your issue goes away.

```java
realm.executeTransactionAsync(new Realm.Transaction() {
    @Override
    public void execute(Realm bgRealm) {
    	bgRealm.where(Task.class).findAll();
    }
}
```

```java
realm.executeTransaction(new Realm.Transaction() {
    @Override
    public void execute(Realm bgRealm) {
    	bgRealm.where(Task.class).findAllAsync();
    }
}
```

**Comment by cmelchior** [comment]
I have been looking at this, and I am still not convinced that Realm is doing everything properly. It is a quit complex setup though, so I am still working my way though it. That synchronous writes solve just affirms my suspicion that perhaps the listeners are not getting called correctly.

**Comment by curiousily** [comment]
Hey @stk1m1 and @cmelchior,

Thank you for looking into this. I realise it might be a bit complex setup.

@stk1m1,
Regarding the question about `isLoaded()` we want to get only the first loaded result and stop the observable from emitting as described here [RealmResults#asObservable()](https://realm.io/docs/java/latest/api/io/realm/RealmResults.html#asObservable).

@stk1m1, 
There might be errors in my code but I fail to see where we nest async query within async transaction (or vice versa). All queries stop emiting after the first result is loaded and `onNext` is called. All transactions are async and `onNext` is called only in `OnSuccess` callbacks.

@cmelchior,
I read somewhere (maybe your comment) that calling `Realm#beginTransaction()` will force refresh/sync between all threads? Is this somewhat related to this issue?

A more general question might be: What is the preferred way to write/read from Realm using observables? One can look at the attempt above (provided in the sample) as implementation that tries to answer that. Are there other approaches?

Still, I am quite a newbie within the RxJava realm (get it? 😄 ). Do you see something wrong with the observable wirings?

**Comment by cmelchior** [comment]
After some more digging, this seems to be caused by #1893.

**Comment by curiousily** [comment]
Thanks. Will follow the progress on that one!

**Comment by cmelchior** [comment]
This have been fixed by #2780 which was just merged, and will be part of next release (Either 1.0.2 or 1.1.0).
